non-determinism in tests

### Description of the problem

I'm seeing intermittent failures when running the test suite:

- `test_get_entity_vals` will pass when run on its own but will sometimes fail if run in the same session as `test_get_meg_crosstalk_fpath` and/or `test_get_meg_calibration_fpath`. See "actual results" for the error.
- `test_rm` will also sometimes fail, haven't narrowed down the cause yet.

### Steps to reproduce

`pytest mne_bids/tests/test_path.py -k "test_meg_ or test_get_entity_vals"` is how I'm *sometimes* seeing the failure of `test_get_entity_vals`. So far I've only seen the `test_rm` failures when running the full test file (`pytest mne_bids/tests/test_path.py`).


### Expected results

tests pass

### Actual results

failure of `test_get_entity_vals`:

```
mne_bids/tests/test_path.py:125: in test_get_entity_vals
    deriv_meg_dir.mkdir(parents=True)
/opt/mambaforge/envs/mnedev/lib/python3.12/pathlib.py:1311: in mkdir
    os.mkdir(self, mode)
E   FileExistsError: [Errno 17] File exists: '/tmp/pytest-of-drmccloy/pytest-12/mnebids_utils_test_bids_ds0/derivatives/pipeline/sub-deriv/ses-deriv/meg'
```

failure of `test_rm`:

```
mne_bids/tests/test_path.py:325: in test_rm
    assert set(captured2.splitlines()) == set(expected)
E   AssertionError: assert {'', '/tmp/py...ts.json', ...} == {'', '/tmp/py...em.json', ...}
E     
E     Extra items in the right set:
E     '/tmp/pytest-of-drmccloy/pytest-0/test_rm0/mnebids_utils_test_bids_ds/participants.tsv'
E     '/tmp/pytest-of-drmccloy/pytest-0/test_rm0/mnebids_utils_test_bids_ds/sub-01'
E     Use -v to get more diff
```

### Additional information

```console
$ pytest --version
pytest 8.3.2
$ mne sys_info
Platform             Linux-6.8.0-45-generic-x86_64-with-glibc2.39
Python               3.12.5 | packaged by conda-forge | (main, Aug  8 2024, 18:36:51) [GCC 12.4.0]
Executable           /opt/mambaforge/envs/mnedev/bin/python3.12
CPU                  Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz (16 cores)
Memory               125.7 GiB

Core
├☒ mne               1.8.0.dev146+g9e932eb38 (outdated, release 1.8.0 is available!)
├☑ numpy             1.26.4 (OpenBLAS 0.3.23.dev with 16 threads)
├☑ scipy             1.14.0
└☑ matplotlib        3.9.2 (backend=qtagg)

Numerical (optional)
├☑ sklearn           1.5.1
├☑ numba             0.60.0
├☑ nibabel           5.2.1
├☑ nilearn           0.10.4
├☑ dipy              1.9.0
├☑ openmeeg          2.5.12
├☑ pandas            2.2.2
├☑ h5io              0.2.4
├☑ h5py              3.11.0
└☐ unavailable       cupy

Visualization (optional)
├☑ pyvista           0.44.1 (OpenGL 4.5.0 NVIDIA 550.120 via NVIDIA GeForce RTX 2060/PCIe/SSE2)
├☑ pyvistaqt         0.11.1
├☑ vtk               9.3.1
├☑ qtpy              2.4.1 (PyQt6=6.7.1)
├☑ ipympl            0.9.4
├☑ pyqtgraph         0.13.7
├☑ mne-qt-browser    0.7.0.dev9+g547b976
├☑ ipywidgets        8.1.3
├☑ trame_client      3.2.3
├☑ trame_server      3.0.3
├☑ trame_vtk         2.8.10
└☑ trame_vuetify     2.6.2

Ecosystem (optional)
├☑ mne-bids          0.16.0.dev23+g20a4ce2d
├☑ mne-nirs          0.8.0.dev5+g7f4e5148
├☑ mne-connectivity  1.8.0.dev146+g9e932eb38
├☑ mne-bids-pipeline 1.10.0.dev15+ga0fc509
├☑ neo               0.13.2
├☑ eeglabio          0.0.2-4
├☑ edfio             0.4.3
├☑ mffpy             0.9.0
├☑ pybv              0.7.5
└☐ unavailable       mne-features, mne-icalabel
```

--- a/doc/whats_new.rst
+++ b/doc/whats_new.rst
@@ -22,6 +22,7 @@ The following authors contributed for the first time. Thank you so much! 🤩
 The following authors had contributed before. Thank you for sticking around! 🤘
 
 * `Stefan Appelhoff`_
+* `Daniel McCloy`_
 
 Detailed list of changes
 ~~~~~~~~~~~~~~~~~~~~~~~~
@@ -50,7 +51,7 @@ Detailed list of changes
 ⚕️ Code health
 ^^^^^^^^^^^^^^
 
-- Nothing yet
+- Tests that were adding or deleting files to/from a session-scoped dataset now properly clean up after themselves, by `Daniel McCloy`_ (:gh:`1347`)
 
 :doc:`Find out what was new in previous releases <whats_new_previous_releases>`
 

--- a/mne_bids/tests/test_path.py
+++ b/mne_bids/tests/test_path.py
@@ -1037,6 +1037,7 @@ def test_match(return_bids_test_dir):
 
     assert bids_path_01.match(check=True) == []
     assert bids_path_01.match(check=False)[0].fpath.name == "sub-01_foo.eeg"
+    bids_path_01.fpath.unlink()  # clean up created file
 
 
 @testing.requires_testing_data
@@ -1118,6 +1119,7 @@ def test_find_matching_paths(return_bids_test_dir):
         check=False,
     )
     assert paths_match == paths_find
+    bids_path_01.fpath.unlink()  # clean up created file
 
 
 @pytest.mark.filterwarnings(warning_str["meas_date_set_to_none"])

--- a/mne_bids/tests/test_path.py
+++ b/mne_bids/tests/test_path.py
@@ -269,8 +269,7 @@ def test_rm(return_bids_test_dir, capsys, tmp_path):
     bids_root = tmp_path / "mnebids_utils_test_bids_ds"
     shutil.copytree(return_bids_test_dir, bids_root)
 
-    # without providing all the entities, ambiguous when trying
-    # to use fpath
+    # without providing all the entities, ambiguous when trying to use fpath
     bids_path = BIDSPath(
         subject=subject_id,
         session=session_id,

--- a/mne_bids/tests/test_path.py
+++ b/mne_bids/tests/test_path.py
@@ -263,10 +263,10 @@ def test_make_folders(tmp_path):
 
 
 @testing.requires_testing_data
-def test_rm(return_bids_test_dir, capsys, tmp_path_factory):
+def test_rm(return_bids_test_dir, capsys, tmp_path):
     """Test BIDSPath's rm method to remove files."""
     # for some reason, mne's logger can't be captured by caplog....
-    bids_root = str(tmp_path_factory.mktemp("test_rm") / "mnebids_utils_test_bids_ds")
+    bids_root = tmp_path / "mnebids_utils_test_bids_ds"
     shutil.copytree(return_bids_test_dir, bids_root)
 
     # without providing all the entities, ambiguous when trying

--- a/mne_bids/tests/test_path.py
+++ b/mne_bids/tests/test_path.py
@@ -494,13 +494,13 @@ def test_find_matching_sidecar(return_bids_test_dir, tmp_path):
     expected_file = op.join("sub-01", "ses-01", "meg", "sub-01_ses-01_coordsystem.json")
     assert str(sidecar_fname).endswith(expected_file)
 
-    # Find multiple sidecars, tied in score, triggering an error
+    # create a duplicate sidecar, which will be tied in match score, triggering an error
+    dupe = Path(str(sidecar_fname).replace("coordsystem.json", "2coordsystem.json"))
+    dupe.touch()
     with pytest.raises(RuntimeError, match="Expected to find a single"):
-        open(
-            str(sidecar_fname).replace("coordsystem.json", "2coordsystem.json"), "w"
-        ).close()
         print_dir_tree(bids_root)
         bids_path.find_matching_sidecar(suffix="coordsystem", extension=".json")
+    dupe.unlink()  # clean up extra file
 
     # Find nothing and raise.
     with pytest.raises(RuntimeError, match="Did not find any"):

--- a/mne_bids/tests/test_path.py
+++ b/mne_bids/tests/test_path.py
@@ -1358,7 +1358,7 @@ def test_bids_path_label_vs_index_entity():
 
 
 @testing.requires_testing_data
-def test_meg_calibration_fpath(return_bids_test_dir):
+def test_meg_calibration_fpath(return_bids_test_dir, tmp_path):
     """Test BIDSPath.meg_calibration_fpath."""
     bids_root = return_bids_test_dir
 
@@ -1382,15 +1382,18 @@ def test_meg_calibration_fpath(return_bids_test_dir):
     with pytest.raises(ValueError, match="Can only find .* for MEG"):
         bids_path_.meg_calibration_fpath
 
-    # Delete the fine-calibration file. BIDSPath.meg_calibration_fpath
-    # should then return None.
+    # Move the fine-calibration file. BIDSPath.meg_calibration_fpath should then be None
     bids_path_ = _bids_path.copy().update(subject="01", root=bids_root)
-    Path(bids_path_.meg_calibration_fpath).unlink()
+    src = Path(bids_path_.meg_calibration_fpath)
+    src.rename(tmp_path / src.name)
     assert bids_path_.meg_calibration_fpath is None
+    # restore the file
+    (tmp_path / src.name).rename(src)
+    assert bids_path_.meg_calibration_fpath is not None
 
 
 @testing.requires_testing_data
-def test_meg_crosstalk_fpath(return_bids_test_dir):
+def test_meg_crosstalk_fpath(return_bids_test_dir, tmp_path):
     """Test BIDSPath.meg_crosstalk_fpath."""
     bids_root = return_bids_test_dir
 
@@ -1414,11 +1417,14 @@ def test_meg_crosstalk_fpath(return_bids_test_dir):
     with pytest.raises(ValueError, match="Can only find .* for MEG"):
         bids_path.meg_crosstalk_fpath
 
-    # Delete the crosstalk file. BIDSPath.meg_crosstalk_fpath should then
-    # return None.
+    # Move the crosstalk file. BIDSPath.meg_crosstalk_fpath should then be None.
     bids_path = _bids_path.copy().update(subject="01", root=bids_root)
-    Path(bids_path.meg_crosstalk_fpath).unlink()
+    src = Path(bids_path.meg_crosstalk_fpath)
+    src.rename(tmp_path / src.name)
     assert bids_path.meg_crosstalk_fpath is None
+    # restore the file
+    (tmp_path / src.name).rename(src)
+    assert bids_path.meg_crosstalk_fpath is not None
 
 
 @testing.requires_testing_data

